### PR TITLE
Fix issue when graphql is true and include permission is WILDCARD

### LINKED ISSUE #1434: Getting Access forbidden message when trying to use filter on graphql queries
### Discussed in https://github.com/Azure/data-api-builder/discussions/1423

<div type='discussions-op-text'>

<sup>Originally posted by **sumchans** April  7, 2023</sup>
I have been trying out the data api builder for azure cosmos db and so far was successfully able to pull data from Azure Cosmos db. The issue I am having now is, when I use filter on the queries I get "Access forbidden to a field referenced in the filter." I believe this has something to do configuring the permissions on the dab-config.json, but I am not able to get a head start so far on this.

Here is my query -
```
query filteritems {
    items(filter: { userPhoneNumber: { contains: "4160000000" } }) {
        items {
            partitionKey
            userPhoneNumber
            userDisplayName
        }
    }
}
```

Here is the result -
```
{
  "errors": [
    {
      "message": "Access forbidden to a field referenced in the filter.",
      "extensions": {
        "code": "AuthorizationCheckFailed"
      }
    }
  ]
}
```

Also attaching my dab-config json file -
```
{
  "$schema": "https://github.com/Azure/data-api-builder/releases/download/v0.6.13/dab.draft.schema.json",
  "data-source": {
    "database-type": "cosmosdb_nosql",
    "options": {
      "database": "Database",
      "schema": "schema.gql"
    },
    "connection-string": "connection-string;"
  },
  "runtime": {
    "graphql": {
      "allow-introspection": true,
      "enabled": true,
      "path": "/graphql"
    },
    "host": {
      "mode": "development",
      "cors": {
        "origins": [],
        "allow-credentials": false
      },
      "authentication": {
        "provider": "StaticWebApps"
      }
    }
  },
  "entities": {
    "Item": {
      "source": "Items",
      "graphql": true,
      "permissions": [
        {
          "role": "anonymous",
          "actions": [
            "*"
          ],
          "fields": {
            "include": [
              "*"
            ]
          }
        }
      ]
    }
  }
}
```
</div>

i have tried  changing the configation and types of permissions but it looks like something broken between the current build (0.6.13)
and the previous (0.35.5). 

the stacktrace shows

```
[dataApi] info: Microsoft.AspNetCore.Hosting.Diagnostics[1]
[dataApi]       Request starting HTTP/1.1 POST http://localhost:4280/graphql application/json 247
[dataApi] info: Microsoft.AspNetCore.Routing.EndpointMiddleware[0]
[dataApi]       Executing endpoint 'Hot Chocolate GraphQL Pipeline'
[dataApi] fail: Azure.DataApiBuilder.Service.Startup[0]
[dataApi]       Access forbidden to a field referenced in the filter.
[dataApi] fail: Azure.DataApiBuilder.Service.Startup[0]
[dataApi]          at Azure.DataApiBuilder.Service.Models.GQLFilterParser.Parse(IMiddlewareContext ctx, IInputField filterArgumentSchema, List`1 fields, BaseQueryStructure queryStructure) in /_/src/Service/Models/GraphQLFilterParsers.cs:line 159   
[dataApi]          at Azure.DataApiBuilder.Service.Resolvers.CosmosQueryStructure.Init(IDictionary`2 queryParams) in /_/src/Service/Resolvers/CosmosQueryStructure.cs:line 160
[dataApi]          at Azure.DataApiBuilder.Service.Resolvers.CosmosQueryEngine.ExecuteAsync(IMiddlewareContext context, IDictionary`2 parameters) in /_/src/Service/Resolvers/CosmosQueryEngine.cs:line 64
[dataApi]          at Azure.DataApiBuilder.Service.Services.ResolverMiddleware.InvokeAsync(IMiddlewareContext context) in /_/src/Service/Services/ResolverMiddleware.cs:line 95
[dataApi]          at HotChocolate.Utilities.MiddlewareCompiler`1.ExpressionHelper.AwaitTaskHelper(Task task)
[dataApi]          at HotChocolate.Execution.Processing.Tasks.ResolverTask.ExecuteResolverPipelineAsync(CancellationToken cancellationToken)
[dataApi]          at HotChocolate.Execution.Processing.Tasks.ResolverTask.TryExecuteAsync(CancellationToken cancellationToken)
[dataApi] info: Microsoft.AspNetCore.Routing.EndpointMiddleware[1]
[dataApi]       Executed endpoint 'Hot Chocolate GraphQL Pipeline'
[dataApi] info: Microsoft.AspNetCore.Hosting.Diagnostics[2]
[dataApi]       Request finished HTTP/1.1 POST http://localhost:4280/graphql application/json 247 - 500 - application/json;+charset=utf-8 864.3125ms
```

the code is not picking up any of the permissions (not explicit, not * and not if left empty) always returning false from  this line

https://github.com/Azure/data-api-builder/blob/893f5039432087ab382deef81ff83bf4380df393/src/Service/Models/GraphQLFilterParsers.cs#L151

if i revert to previous version, the exact same definition works fine.

### PR DESCRIPTION
**Pull request by tarazou9**
## Why make this change?

- Closes #1434 
  - When ```graphql``` is set to true without singular type specified in runtime config file, in addition to when field auth ```include``` permission is using WILDCARD, this was giving user's access forbidden error.

## What is this change?

- Reason for causing this is because ```GraphQLSingularTypeToEntityNameMap``` in ```RuntimeConfig.cs``` is not adding the entity type when there is no singular type passed in, and for WILDCARD permission check, the ```AuthorizationResolver``` is retrieving all the fields columns using the ```GraphQLSingularTypeToEntityNameMap``` keys.

## How was this tested?

- [x] Integration Tests

### CODE REVIEW

File: config-generators/cosmosdb_nosql-commands.txt [hunk 1]
```diff
@@ -12,4 +12,6 @@ add Earth --config "dab-config.CosmosDb_NoSql.json" --source "graphqldb.earth" -
 update Earth --config "dab-config.CosmosDb_NoSql.json" --permissions "anonymous:create" --fields.include "id" --fields.exclude "name"
 update Earth --config "dab-config.CosmosDb_NoSql.json" --permissions "anonymous:read" --fields.include "id,type" --fields.exclude "name"
 update Earth --config "dab-config.CosmosDb_NoSql.json" --permissions "anonymous:update" --fields.exclude "*"
-update Earth --config "dab-config.CosmosDb_NoSql.json" --permissions "authenticated:create,read,update,delete"
\ No newline at end of file
+update Earth --config "dab-config.CosmosDb_NoSql.json" --permissions "authenticated:create,read,update,delete"
+add Sun --config "dab-config.CosmosDb_NoSql.json" --source "graphqldb.sun" --permissions "anonymous:create,update,delete" --graphql true
+update Sun --config "dab-config.CosmosDb_NoSql.json" --permissions "anonymous:read" --fields.exclude "*"
```
Comment (abhishekkumams) [inline config-generators/cosmosdb_nosql-commands.txt:17]: here, the command is used to add `fields.exclude`, but in the config file it's showing under `fields.include`

File: src/Service.Tests/dab-config.CosmosDb_NoSql.json [hunk 2]
```diff
@@ -148,6 +148,28 @@
...
+                  "*"
+                ]
+              }
+            },
+            "update",
+            "delete"
+          ]
+      }
+      ],
+      "graphql": true
```
Comment (mbhaskar) [inline src/Service.Tests/dab-config.CosmosDb_NoSql.json:171]: Can you double check if there are any other combinations or possibilities that needs to be tested here?

File: config-generators/cosmosdb_nosql-commands.txt (hunk 1 above)
Comment (tarazou9) [reply config-generators/cosmosdb_nosql-commands.txt]: Yes, this caused the integration tests to fail as well. I've fixed this, thank you for catching it!

File: src/Service.Tests/CosmosTests/QueryFilterTests.cs [hunk 3]
```diff
@@ -834,6 +834,28 @@ public async Task TestQueryFieldAuthConflictingWithFilterFieldAuth_Unauthorized(
             string errorMessage = response.ToString();
             Assert.IsTrue(errorMessage.Contains("The current user is not authorized to access this resource."));
         }
+
+        /// <summary>
+        /// Tests that the field level query filter succeeds requests
+        /// when GraphQL is set to ture without setting singular type in runtime config and
```
Comment (seantleonard) [inline src/Service.Tests/CosmosTests/QueryFilterTests.cs:840]:
```suggestion
        /// when GraphQL is set to true without setting singular type in runtime config and
```

File: src/Service.Tests/dab-config.CosmosDb_NoSql.json [hunk 4]
```diff
@@ -148,6 +148,28 @@
...
+          "actions": [
+            "create",
+            {
+              "action": "read",
+              "fields": {
+                "include": [
+                  "*"
+                ]
+              }
+            },
```
Comment (seantleonard) [inline src/Service.Tests/dab-config.CosmosDb_NoSql.json:165]: This section (which is valid) can also be represented as JSON string "read" which implies that all fields are included. That should behave the same way.

File: config-generators/cosmosdb_nosql-commands.txt [hunk 5]
```diff
@@ -12,4 +12,6 @@ add Earth --config "dab-config.CosmosDb_NoSql.json" --source "graphqldb.earth" -
 update Earth --config "dab-config.CosmosDb_NoSql.json" --permissions "anonymous:create" --fields.include "id" --fields.exclude "name"
 update Earth --config "dab-config.CosmosDb_NoSql.json" --permissions "anonymous:read" --fields.include "id,type" --fields.exclude "name"
 update Earth --config "dab-config.CosmosDb_NoSql.json" --permissions "anonymous:update" --fields.exclude "*"
-update Earth --config "dab-config.CosmosDb_NoSql.json" --permissions "authenticated:create,read,update,delete"
\ No newline at end of file
+update Earth --config "dab-config.CosmosDb_NoSql.json" --permissions "authenticated:create,read,update,delete"
+add Sun --config "dab-config.CosmosDb_NoSql.json" --source "graphqldb.sun" --permissions "anonymous:create,update,delete" --graphql true
+update Sun --config "dab-config.CosmosDb_NoSql.json" --permissions "anonymous:read" --fields.include "*"
```
Comment (aaronpowell) [inline config-generators/cosmosdb_nosql-commands.txt:17]: Make sure you drop a new line on the end of this file as if you don't the last line isn't included on Linux/MacOS.

**Comment by aaronpowell** [comment]
I feel like #1402 will tackle this as it is more _aggressive_ in the way that it sets the singular name, resulting in the `RuntimeConfig` always having a defined singular name, rather than having to rely on additional mapping data structures.

File: src/Service.Tests/CosmosTests/QueryFilterTests.cs (hunk 3 above)
Comment (tarazou9) [reply src/Service.Tests/CosmosTests/QueryFilterTests.cs]: fixed.

File: src/Service.Tests/CosmosTests/QueryFilterTests.cs [hunk 6]
```diff
@@ -834,6 +834,28 @@ public async Task TestQueryFieldAuthConflictingWithFilterFieldAuth_Unauthorized(
...
+        /// all the columns are able to be retrieved for authorization validation. 
+        /// </summary>
+        [TestMethod]
+        public async Task TestQueryFilterFieldAuthWithoutSingularType()
+        {
+            string gqlQuery = @"{
+                suns(first: 1, " + QueryBuilder.FILTER_FIELD_NAME + @" : {id : {eq : """ + _idList[0] + @"""}})
+                { 
+                    items {
+                        id
```
Comment (Aniruddh25) [inline src/Service.Tests/CosmosTests/QueryFilterTests.cs:851]: can we retrieve `name` too to test that `*`/Wildcard evaluates to all columns?

File: src/Service.Tests/CosmosTests/QueryFilterTests.cs [hunk 7]
```diff
@@ -36,7 +36,7 @@ public static void TestFixtureSetup(TestContext context)
             OverrideEntityContainer("Planet", _containerName);
             OverrideEntityContainer("Earth", _containerName);
             OverrideEntityContainer("StarAlias", _containerName);
-            OverrideEntityContainer("TagAlias", _containerName);
```
Comment (Aniruddh25) [inline src/Service.Tests/CosmosTests/QueryFilterTests.cs:39]: Is `TagAlias` no longer used? If yes, did we remove from all the occurrences?

Comment (tarazou9) [reply src/Service.Tests/CosmosTests/QueryFilterTests.cs:39]: Yes, it's used here. https://github.com/Azure/data-api-builder/blob/a52db81da1d9a2248f1cf83b978a58dae5290dda/src/Service.Tests/CosmosTests/QueryFilterTests.cs#L665

**Review by Aniruddh25 — APPROVED** [review]
LGTM, thanks for the fix and testing all other scenarios!

File: config-generators/cosmosdb_nosql-commands.txt [hunk 8]
```diff
@@ -12,4 +12,6 @@ add Earth --config "dab-config.CosmosDb_NoSql.json" --source "graphqldb.earth" -
 update Earth --config "dab-config.CosmosDb_NoSql.json" --permissions "anonymous:create" --fields.include "id" --fields.exclude "name"
 update Earth --config "dab-config.CosmosDb_NoSql.json" --permissions "anonymous:read" --fields.include "id,type" --fields.exclude "name"
```
Comment (abhishekkumams) [inline config-generators/cosmosdb_nosql-commands.txt:13]: can you try once with include `all` and exclude `one`, for example: --include "*" and --exclude "name"

Comment (tarazou9) [reply config-generators/cosmosdb_nosql-commands.txt:13]: @abhishekkumams Added test for verifying this.

**Review by seantleonard — APPROVED** [review]

**Pull request merged**